From 50ed1c570a36ceeb61bb4a0ce0e1240af35f3da2 Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Thu, 09 Jun 2022 19:51:52 +0300
Subject: [PATCH] [SW-70527] gaudi2: ras: sob_xdma sram rings repair

This patch added support for BIRA, BISR and BIST operations
for few Sync manager and xDMA(PDMA and KDMA) SRAM rings.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: Icc66763a9be31671cc9b2ddc74d772793fa30ff6
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index cb8ad1d..685e9ac 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -38,6 +38,7 @@
 #define DELAY_4K_USEC			4000
 #define DELAY_20_USEC			20
 #define DELAY_104_USEC			104
+#define DELAY_36_USEC			36
 
 #define GAUDI2_DFX_NUM			5
 #define GAUDI2_MAX_RING_NUM		28
@@ -464,6 +465,58 @@
 	}
 }
 
+static void gaudi2_sob_xdma_ring_pattern(enum gaudi2_sram_self_repair_op op)
+{
+	uint32_t expected_ones, expected_zeros;
+
+	regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
+	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_5K_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_36_USEC);
+	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
+		udelay(DELAY_5K_USEC);
+	} else {
+		preboot_puts("Invalid sram repair operation\n");
+		return;
+	}
+	regmap_writel(GAUDI2_RING_PATTERN_VAL39, mmPSOC_DFT_APB_WORD0);
+	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000010100000010100000010100;
+	expected_zeros = 0b00000000001011100001011100001011;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST)
+		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
+	else
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+	expected_ones = 0b00000000000001010000001010000000;
+	expected_zeros = 0b00000000010000101110000101110000;
+	gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+
+	if (op == SRAM_SELF_REPAIR_OP_BIST) {
+		regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
+
+		regmap_writel(GAUDI2_RING_PATTERN_VAL40, mmPSOC_DFT_APB_WORD0);
+		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
+
+		expected_ones = 0b00000000000000000000000000000000;
+		expected_zeros = 0b00000000000000010010010010010000;
+		gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
+	}
+
+}
+
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
 {
 	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
@@ -507,7 +560,7 @@
 				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
 				[SRAM_RING_TYPE_XBAR_MID] = gaudi2_xbar_mid_ring_pattern,
 				[SRAM_RING_TYPE_VDEC] = gaudi2_vdec_ring_pattern,
-				[SRAM_RING_TYPE_SOB_XDMA] = NULL,
+				[SRAM_RING_TYPE_SOB_XDMA] = gaudi2_sob_xdma_ring_pattern,
 				[SRAM_RING_TYPE_NIC] = NULL,
 				[SRAM_RING_TYPE_MME] = NULL,
 				[SRAM_RING_TYPE_HMMU_XBAR] = NULL,
@@ -530,7 +583,8 @@
 				ring_type == SRAM_RING_TYPE_TPC ||
 				ring_type == SRAM_RING_TYPE_SOB_ROT ||
 				ring_type == SRAM_RING_TYPE_XBAR_MID ||
-				ring_type == SRAM_RING_TYPE_VDEC) {
+				ring_type == SRAM_RING_TYPE_VDEC ||
+				ring_type == SRAM_RING_TYPE_SOB_XDMA) {
 				gaudi2_sram_id_sel();
 				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
 				gaudi2_sram_region_ring_pattern[ring_type](op);
diff --git a/arch/gaudi2/common/common.h b/arch/gaudi2/common/common.h
index e35e14e..6ee429f 100644
--- a/arch/gaudi2/common/common.h
+++ b/arch/gaudi2/common/common.h
@@ -260,6 +260,9 @@
 #define GAUDI2_RING_PATTERN_VAL35		0x100002a
 #define GAUDI2_RING_PATTERN_VAL36		0x100002e
 #define GAUDI2_RING_PATTERN_VAL37		0x1000016
+#define GAUDI2_RING_PATTERN_VAL38		0x1000032
+#define GAUDI2_RING_PATTERN_VAL39		0x1000037
+#define GAUDI2_RING_PATTERN_VAL40		0x1000019
 
 
 void gaudi2_axi_drain_ena(void);
